usecases/internal/measurement.go

```
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+	for i := range measurements {
+		ptrMeasurements[i] = &measurements[i]
+	}
+	
+	valid, err := validator.ValidateFirst(ptrMeasurements)
+	if err != nil {
+		return 0, api.ErrDataNotAvailable
+	}
+	
+	if valid == nil || valid.Value == nil {
+		return 0, api.ErrDataNotAvailable
+	}
+	
+	return valid.Value.GetValue(), nil
+}
+
+// Measurement-specific validation rules
+
+// RequireMeasurementId ensures MeasurementId is present.
+//
+// This is the most basic measurement validation rule.
+// Almost all measurement validators should include this rule.
+func RequireMeasurementId() ValidationRule[*model.MeasurementDataType] {
+	return RequireField(
+		func(m *model.MeasurementDataType) *model.MeasurementIdType { return m.MeasurementId },
+		"MeasurementId",
+	)
+}
+
+// RequireMeasurementValue ensures Value is present.
+//
+// This rule ensures that measurement data contains an actual value.
+// Use this when you need to extract numeric values from measurements.
+func RequireMeasurementValue() ValidationRule[*model.MeasurementDataType] {
+	return RequireScaledNumber(
+		func(m *model.MeasurementDataType) *model.ScaledNumberType { return m.Value },
+		"Value",
+	)
+}
+
+// RequireValueType ensures ValueType matches expected type
+func RequireValueType(expected model.MeasurementValueTypeType) ValidationRule[*model.MeasurementDataType] {
+	return func(m *model.MeasurementDataType) error {
+		if expected == "" {
+			return nil // Skip if no expected type specified
+		}
+		if m.ValueType == nil {
+			return fmt.Errorf("ValueType is required to be %s", expected)
+		}
+		if *m.ValueType != expected {
+			return fmt.Errorf("ValueType must be %s, got %s", expected, *m.ValueType)
+		}
+		return nil
+	}
+}
+
+// RequireValueSource ensures ValueSource is one of allowed types
+func RequireValueSource(allowed ...model.MeasurementValueSourceType) ValidationRule[*model.MeasurementDataType] {
+	if len(allowed) == 0 {
+		return func(m *model.MeasurementDataType) error { return nil }
+	}
```

This feels like a caller error that we shouldn't be silently ignoring here.